azurelinuxagent/ga/exthandlers.py

```diff
@@ -476,20 +479,45 @@ def handle_enable(self, ext_handler_i):
             if old_ext_handler_i is None:
                 ext_handler_i.install()
             elif ext_handler_i.version_ne(old_ext_handler_i):
-                old_ext_handler_i.disable()
-                ext_handler_i.copy_status_files(old_ext_handler_i)
-                if ext_handler_i.version_gt(old_ext_handler_i):
-                    ext_handler_i.update()
-                else:
-                    old_ext_handler_i.update(version=ext_handler_i.ext_handler.properties.version)
-                old_ext_handler_i.uninstall()
-                old_ext_handler_i.remove_ext_handler()
-                ext_handler_i.update_with_install()
+                ExtHandlersHandler._update_extension_handler(old_ext_handler_i, ext_handler_i)
         else:
             ext_handler_i.update_settings()
 
         ext_handler_i.enable()
 
+    @staticmethod
+    def _update_extension_handler(old_ext_handler_i, ext_handler_i):
+
+        def execute_old_handler_command(ext_handler_i, op):
+            """
+            Created a common function to execute all commands that need to be executed from the old handler
+            so that it can have a common exception handling mechanism
+            :param ext_handler_i: The extension handler instance to run the operation on
+            :param op: Which operation to run
+            """
+            try:
+                if op == WALAEventOperation.Disable:
+                    ext_handler_i.disable()
+                elif op == WALAEventOperation.UnInstall:
+                    # Currently uninstall failures swallows the exception so we wont catch anything here
+                    ext_handler_i.uninstall()
+            except ExtensionError as e:
+                # Reporting the event with the old handler and raising a new ExtensionUpdateError to set the
+                # handler status on the new version
+                msg = ustr(e)
+                ext_handler_i.report_event(message=msg, is_success=False)
+                raise ExtensionUpdateError(msg)
+
+        execute_old_handler_command(old_ext_handler_i, WALAEventOperation.Disable)
+        ext_handler_i.copy_status_files(old_ext_handler_i)
+        if ext_handler_i.version_gt(old_ext_handler_i):
+            ext_handler_i.update()
+        else:
+            old_ext_handler_i.update(version=ext_handler_i.ext_handler.properties.version)
```

Comment:
wow! we support extension downgrade?

Reply:
Haha yes, this is at par with Windows. Not sure how much its been tested though.

Reply:
would you open a work item to test it and double-check whether the extension publishers doc talks about this? i'm sure some extensions publishers will be surprised (let's also check whether the VERSION env variable is documented). Thanks!

Reply:
Added a [task](https://msazure.visualstudio.com/One/_workitems/edit/5209494). Will follow up later about this